src/main/java/io/cdap/plugin/batch/sink/RecordToBSONWritableTransformer.java

```diff
@@ -0,0 +1,294 @@
+    recordWithIdFields.addAll(copiedFields);
+
+    String recordName = Objects.requireNonNull(original.getSchema().getRecordName(), "Record name can not be empty");
+    Schema recordWithIdSchema = Schema.recordOf(recordName, recordWithIdFields);
+    StructuredRecord.Builder builder = StructuredRecord.builder(recordWithIdSchema);
+    for (Schema.Field field : original.getSchema().getFields()) {
+      if (idFieldName.equals(field.getName())) {
+        builder.set(DEFAULT_ID_FIELD_NAME, original.get(idFieldName));
+        continue;
+      }
+
+      builder.set(field.getName(), original.get(field.getName()));
+    }
+
+    return builder.build();
+  }
+
+
+  private DBObject extractRecord(Object value, @Nullable String idFieldName) {
+    if (value == null) {
+      // Return 'null' value as it is
+      return null;
+    }
+    StructuredRecord record = (StructuredRecord) value;
+    List<Schema.Field> fields = Objects.requireNonNull(record.getSchema().getFields(), "Schema fields cannot be empty");
+    BasicDBObjectBuilder bsonBuilder = BasicDBObjectBuilder.start();
+    for (Schema.Field field : fields) {
+      Schema.LogicalType fieldLogicalType = field.getSchema().isNullable()
+        ? field.getSchema().getNonNullable().getLogicalType()
+        : field.getSchema().getLogicalType();
+
+      if (!Strings.isNullOrEmpty(idFieldName) && idFieldName.equals(field.getName())) {
+        bsonBuilder.add(idFieldName, extractObjectId(record.get(field.getName())));
+        continue;
+      }
+
+      if (fieldLogicalType == null) {
+        bsonBuilder.add(field.getName(), extractValue(record.get(field.getName()), field.getSchema()));
+        continue;
+      }
+
+      // Get values of logical types properly
+      switch (fieldLogicalType) {
+        case TIMESTAMP_MILLIS:
+        case TIMESTAMP_MICROS:
+          ZonedDateTime timestamp = Objects.requireNonNull(record.getTimestamp(field.getName()));
+          bsonBuilder.add(field.getName(), Date.from(timestamp.toInstant()));
+          break;
+        case TIME_MILLIS:
+        case TIME_MICROS:
+          LocalTime time = Objects.requireNonNull(record.getTime(field.getName()));
+          bsonBuilder.add(field.getName(), time.toString());
+          break;
+        case DATE:
+          LocalDate localDate = Objects.requireNonNull(record.getDate(field.getName()));
+          Date date = Date.from(localDate.atStartOfDay(ZoneId.ofOffset("UTC", ZoneOffset.UTC)).toInstant());
+          bsonBuilder.add(field.getName(), date);
+          break;
+        case DECIMAL:
+          BigDecimal decimal = Objects.requireNonNull(record.getDecimal(field.getName()));
+          bsonBuilder.add(field.getName(), new Decimal128(decimal));
+          break;
+        default:
+          throw new UnexpectedFormatException("Field type '" + fieldLogicalType + "' is not supported.");
+      }
+    }
+    return bsonBuilder.get();
+  }
+
+  private ObjectId extractObjectId(Object value) {
```

Comment:
what happens if these values are invalid ids? Need to make sure that whatever gets logged makes it clear what the problem is.

Reply:
In this case, `IllegalArgumentException` with an error message will be thrown. Changed to wrap this into `UnexpectedFormatException`:
```
...
    } catch (IllegalArgumentException e) {
      throw new UnexpectedFormatException(
        String.format("Invalid value '%s' of the identifier field '%s'.", value, fieldName), e);
    }
...
```